### DISCUSSION TITLE
For workerDirectory when using msw in workspace used within a certain npm workspaces.

### DISCUSSION THREAD

**Discussion by k35o** [mswjs/msw#2065] [Q&A] [answered]
Suppose `npm workspaces` is constructed as follows.
```
- root(package.json)
  - workspaceA(package.json)
  - workspaceB(package.json)
```
And assume that `msw` is used only within `workspaceA`.

```json:workspaceA/package.json
// workspaceA/package.json
{
  ...,
  "devDependencies": {
    ...,
    "msw": "2.0.0",
    ...,
  },
  ...,
  "msw": {
    "workerDirectory": "./public"
  }
}
```
Running `npm i -D msw@latest -w workspaceA` in this situation will not update `mockServiceWorker.js` to see `package.json` in the `root` directory.
https://github.com/mswjs/msw/blob/b3e47f71d3a0333aaeab3da5839cb9068a30d7ef/config/scripts/postinstall.js#L8

How can I configure `workspaceA/package.json` for updating `mockServiceWorker.js`? 
I don't want to write `msw.workerDirectory` in the `root` `package.json`, Because I don't want to leak `workspaceA` information out.

**Comment by kettanaito** [comment, marked as answer]
Hi, @k35o. 

The `msw.workerDirectory` must always be specified in the root-level `package.json`. That's the CWD you have when you run `npm install`. MSW cannot and mustn't assume workspaces or `-w` flags. 

Here's what you should do instead.

1. `npm i msw@latest -D -w packageA`
2. `npx msw init ./packages/packageA --save` (from the root).

This will store the public path of your nested package for future MSW updates in the root-level `package.json`:

```json
// ./package.json
{
  "msw": {
    "workerDirectory": ["./packages/packageA/public"]
  }
}
```

You can also keep a list of values in case multiple packages in your monorepo need the worker script. 

## Related materials

- [Managing the worker](https://mswjs.io/docs/best-practices/managing-the-worker)
